examples/explore.py

```diff
@@ -0,0 +1,506 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+"""Run a parallelized exhaustive search of an action space.
+
+All possible combinations of actions up to a finite limit are
+evaluated, but partial sequences of actions that end up in the same
+state are deduplicated, sometimes dramatically reducing the size of
+the search space. Can also be configured to do a beam search.
+
+Example usage:
+
+    $ python explore.py --env=llvm-ic-v0 --benchmark=cBench-v0/dijkstra \
+       --episode_length=10 --actions=-simplifycfg,-instcombine,-mem2reg,-newgvn
+
+Use --help to list the configurable options.
+"""
+import hashlib
+import itertools
+import math
+from enum import IntEnum
+from heapq import nlargest
+from multiprocessing import cpu_count
+from multiprocessing.pool import ThreadPool
+from queue import Queue
+from threading import Lock
+from time import time
+
+import humanize
+from absl import app, flags
+
+from compiler_gym.util.flags.benchmark_from_flags import benchmark_from_flags
+from compiler_gym.util.flags.env_from_flags import env_from_flags
+
+flags.DEFINE_list(
+    "actions",
+    [],
+    "A list of action names to enumerate. If not provided, all actions are used "
+    "(warning: this might take a long time!)",
+)
+flags.DEFINE_integer("episode_length", 5, "The number of steps in each episode.")
+flags.DEFINE_integer(
+    "nproc", cpu_count(), "The number of parallel worker threads to run."
+)
+flags.DEFINE_integer(
+    "topn",
+    0,
+    "If positive, explore only the top n states "
+    "for each sequence length. This is in effect the widt "
+    "of a beam search.",
+)
+flags.DEFINE_integer(
+    "show_topn", 3, "Show this many top sequences " "at each sequence length."
+)
+
+FLAGS = flags.FLAGS
+
+
+class CustomEnv:
+    """A wrapper for an LLVM env that takes a subset of the actions.
+
+    Taking a subset in the env avoids the easy error to make to pass in
+    i as an action instead of actions[i] where actions is the subset.
+    """
+
+    def __init__(self):
+        self._env = env_from_flags(benchmark_from_flags())
+        try:
+            # Project onto the subset of transformations that have
+            # been specified to be used.
+            if not FLAGS.actions:
+                self._action_indices = list(range(len(self._env.action_space.names)))
+            else:
+                self._action_indices = [
+                    self._env.action_space.flags.index(a) for a in FLAGS.actions
+                ]
```

Comment:
Good idea, it could then be tied to a common flag and picked up in env_from_flags() with some nicer error messages etc. than what I did here. I'll add it to my todo.

Reply:
Yep, hooking that into the common flags would be great!